src/runtime/virtcontainers/utils/utils.go

```diff
@@ -306,43 +306,74 @@ func ConvertAddressFamily(family int32) pbTypes.IPFamily {
 	}
 }
 
-// WaitLocalProcess waits for the specified process for up to timeoutSecs seconds.
-//
-// Notes:
-//
-// - If the initial signal is zero, the specified process is assumed to be
-//   attempting to stop itself.
-// - If the initial signal is not zero, it will be sent to the process before
-//   checking if it is running.
-// - If the process has not ended after the timeout value, it will be forcibly killed.
-func WaitLocalProcess(pid int, timeoutSecs uint, initialSignal syscall.Signal, logger *logrus.Entry) error {
-	var err error
+func isProcessDead(pid int) bool {
+	return syscall.Kill(pid, syscall.Signal(0)) != nil
+}
 
-	// Don't support process groups
-	if pid <= 0 {
-		return errors.New("can only wait for a single process")
+func waitUsingPidfd(pid int, timeoutSecs uint, logger *logrus.Entry) (bool, error) {
+	pidfd, err := unix.PidfdOpen(pid, 0)
+
+	if err != nil {
+		if err == unix.ESRCH {
+			return false, nil
+		}
+
+		return true, err
 	}
 
-	if initialSignal != syscall.Signal(0) {
-		if err = syscall.Kill(pid, initialSignal); err != nil {
-			if err == syscall.ESRCH {
-				logger.WithField("pid", pid).Warnf("kill encounters ESRCH, process already finished")
-				return nil
-			}
+	defer unix.Close(pidfd)
 
-			return fmt.Errorf("Failed to send initial signal %v to process %v: %v", initialSignal, pid, err)
+	start := time.Now()
+	var n int
+
+	for {
+		remaining := int(timeoutSecs*1000) - int(time.Since(start).Milliseconds())
+		if remaining < 0 {
+			remaining = 0
+		}
+
+		n, err = unix.Poll([]unix.PollFd{{Fd: int32(pidfd), Events: unix.POLLIN}}, remaining)
+		if err != unix.EINTR {
+			break
 		}
 	}
 
-	pidRunning := true
+	if err != nil || n != 1 {
+		logger.Warnf("process %v still running after waiting %ds", pid, timeoutSecs)
+		return true, err
+	}
+
+	return false, nil
+}
 
+func waitUsingLegacy(pid int, timeoutSecs uint, logger *logrus.Entry) bool {
 	secs := time.Duration(timeoutSecs)
 	timeout := time.After(secs * time.Second)
 
-	// Wait for the VM process to terminate
-outer:
+	errc := make(chan error, 1)
+
+	go func() {
+		p, err := os.FindProcess(pid)
+		if err == nil {
+			_, err = p.Wait()
+		}
+		errc <- err
+	}()
+
```

Comment:
I missed that during the first review but there's a problem here. Let's suppose the hypervisor process is stopped with kill(SIGSTOP), this `p.Wait()` call will block until the process is runnable again, i.e. at some random time. It doesn't look great that `waitUsingLegacy()` might return and leave the goroutine behind... the result of `p.Wait()` would be lost IIUC.

Reply:
Yeah, that goroutine will remain blocked until the process exits, no way to cancel waiting, this a problem in general in linux with waiting for pids.

Reply:
I removed that part, and left wait4 & signal 0.